source/identifier.ts

```diff
@@ -1,14 +1,17 @@
-import {Context as TelegrafContext} from 'telegraf'
+import {Context as TelegrafContext, NarrowedContext} from 'telegraf'
 import {markdown, markdownv2, html} from 'telegram-format'
-import {MessageEntity, Message} from 'typegram'
+import {MessageEntity, Message, Update} from 'typegram'
 
 const URL_TEXT = '\u200C'
 const BASE_URL = 'http://t.me/#'
 const URL_SEPERATOR = '#'
 
-// TODO: refactor when GuardedContext (or whatever its named) is out (see https://github.com/telegraf/telegraf/discussions/1284)
+// ReplyMessage from typegram is not exported
 type ReplyToMessage = NonNullable<Message.CommonMessage['reply_to_message']>
-export type ReplyToMessageContext<Context extends TelegrafContext> = Context & {message: Message.CommonMessage & {reply_to_message: ReplyToMessage}}
+
+type MessageUpdateContext<Context extends TelegrafContext> = NarrowedContext<Context, Update.MessageUpdate>
+export type ReplyToMessageContext<Context extends TelegrafContext> = MessageUpdateContext<Context> & {message: MessageUpdateContext<Context>['message'] & {reply_to_message: ReplyToMessage}}
```

Comment:
```suggestion
type NewNonChannelMessage = Update.MessageUpdate['message']
export type ReplyToMessageContext<C extends TelegrafContext> = NarrowedContext<
	C,
	{
		update_id: number
		message: Pick<NewNonChannelMessage, Message.TextMessage> & {
			reply_to_message: ReplyToMessage
		}
	}
>
```

Trickier than I'd like. https://github.com/KnorpelSenf/typegram/pull/10 would have allowed `Update.MessageUpdate<Message.TextMessage & { reply_to_message: ReplyToMessage }>` @KnorpelSenf. Alternatively,

```ts
type UpdateMap = {
  [T in UpdateType]: Extract<tg.Update, Record<T, object>>[T]
}

type NarrowedUpdate<T extends UpdateType, Q, E = unknown> = T extends unknown
  ? { update_id: number } & Record<T, Extract<UpdateMap[T], Q> & E>
  : never
```

allows

```ts
NarrowedUpdate<
  'message',
  Message.TextMessage,
  { reply_to_message: ReplyToMessage }
>
```

Reply:
`NarrowedContext` is actually pretty nasty internally, I'd suggest to wait until I'll try cleaning it up for 4.3...

Reply:
Simpler interface, needs more testing and is too slow for current `NarrowedContext`:

```ts
type OmitUndefined<T extends object, B = T> = T extends unknown
  ? {
      [K in keyof T as B extends Partial<Record<K, undefined>>
        ? never
        : K]: T[K]
    }
  : never

type NarrowedUpdate<
  T extends tt.UpdateType,
  Q extends object
> = T extends unknown
  ? { update_id: number } & Record<
      T,
      OmitUndefined<Deunionize<UpdateMap[T]> & Q>
    >
  : never
```

```ts
NarrowedUpdate<
  'message',
  { text: string; reply_to_message: ReplyToMessage }
>
```